### PR TITLE
feat: Implement minimum TLS version for clients

### LINKED ISSUE #8124: Support for Advanced TLS configurations for client type input plugins.
Proposal:
Implement [Advanced TLS configuration](https://github.com/influxdata/telegraf/blob/v1.15.3/docs/TLS.md) support for client type input plugin [GNMI input ](https://github.com/influxdata/telegraf/tree/v1.15.3/plugins/inputs/gnmi) (and all plugins) permitting TLS cipher suites/versions to be specified at the client-side.

Current behavior:
Currently, it is only possible to specify tls_cipher_suites/tls_versions on server type input plugins. I believe the current behavior for client type plugins defaults to a specific TLS cipher suite supported by the running version of golang.

Desired behavior:
Allow tls_cipher_suites and tls_min/max_version to be specified in configuration files for client-type input plugins... 

Use case: 
Implementing this would allow users to choose which TLS cipher suites and versions should be used for a given gRPC connection providing a higher level of security/control/compatibility. I assume this would share the same benefits if also implemented with other client type input plugins in addition to [GNMI input](https://github.com/influxdata/telegraf/tree/v1.15.3/plugins/inputs/gnmi).

### LINKED ISSUE #8171: Common plugin TLS module allows for insecure ciphers
The TLS module included for plugins allows insecure versions 1.0 and 1.1.

See the plugins common file: `plugins/common/tls/common.go`

### LINKED ISSUE #8699: TLS stopped working with Telegraf v1.15 and newer.
### Relevant telegraf.conf:

```toml
[global_tags]
[agent]
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "5s"
  flush_interval = "1m"
  flush_jitter = "5s"
  debug = true
  quiet = false
  hostname = "test1"
  omit_hostname = false

[[inputs.jti_openconfig_telemetry]]
  servers = ["router1.mgt.net:50051", "router2.mgt.net:50051"]
    sensors = ["interfaces /junos/system/linecard/interface/","ifaceDesc /interfaces/interface/state/description/","lsps /junos/services/label-switched-path/usage/","cpu /junos/system/linecard/cpu/memory/","npu /junos/system/linecard/npu/memory/"] 
  sample_frequency = "60000ms"
  username = "$routerUser"
  password = "$routerPass"
  client_id = "$containerName"
  enable_tls = true
  tls_ca = "/etc/telegraf/router_ca.pem"
  insecure_skip_verify = false
  str_as_tags = false

[[outputs.file]]
  files = ["stdout"]
```

### System info:

Telegraf v1.17.0
MacOS 11.0.1

### Steps to reproduce:

1. Brew install Telegraf 1.17
2. Run Telegraf using the above config

### Expected behavior:

Telegraf should be able to connect to the routers and start streaming the telemetry data.
### Actual behavior:

Get this error:
[inputs.jti_openconfig_telemetry] Could not initiate login check for nep6bbrj03.mgt.cox.net:50051: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: EOF"
### Additional info:
Downgrade to version 1.14 or older and TLS works.

### PR DESCRIPTION
**Pull request by srebhan**
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #8699
resolves #8171
resolves #8124 (at least partially)

replaces #8959

All credits for this PR go to @jdstrand! This PR allows to set a minimum TLS version required by the client for all plugins that use `common/tls` to configure their TLS setup. This includes, but is probably not limited to, `inputs.http`, `inputs.gnmi` and `inputs.jti_openconfig_telemetry`. This allows to change Golang's default of TLS 1.2 in both directions, i.e. downgrade (which implies some security risk) as well as setting it to TLS 1.3.
Unit-tests are added by the PR to check if lower server versions are rejected successfully.

### CODE REVIEW

File: plugins/inputs/gnmi/sample.conf
Comment (sspaink) [inline plugins/inputs/gnmi/sample.conf:20]:
```suggestion
  ## If unset, the value defaults to Golang's default which is TLS v1.2 currently.
```

File: plugins/inputs/gnmi/README.md
Comment (powersj) [inline plugins/inputs/gnmi/README.md:36]: I don't think we should say in each config file what the minimum TLS version is. It is asking for trouble if/when it changes and we do not catch that.

under -> unset

File: plugins/common/tls/config.go
Comment (jdstrand) [inline plugins/common/tls/config.go:91]: Thanks for picking this up! I've not reviewed the whole PR but thought I'd mention that 1.0 and 1.1 are now widely deprecated (all the browsers, twitter, etc, etc) and so I suggest logging a warning if a) the min version is less than 1.2 or b) the default tls version from Go is less than 1.2[1]. Since telegraf 1.x will (presumably) continue to use the Go defaults, this has the added benefit of discoverability for people upgrading. It will continue to be useful in 2.x when it (presumably) sets the min version to 1.2 since users choosing a lower version will be reminded they are choosing an insecure configuration.

[1] https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/crypto/tls/common.go;l=685 - clients default to 1.2 but servers default to 1.0 so runtime detecting would future-proof this.

File: plugins/inputs/gnmi/README.md
Comment (jdstrand) [reply plugins/inputs/gnmi/README.md:36]: I can see the argument on either side for putting this in each plugin that uses it, however it feels like putting it only in the global `[agent]` section might make the most sense. Eg:
```
[agent]
...
## Minimal TLS version to use for plugins that support TLS.
## If unset, the value defaults to Golang's default which is TLS v1.2 for clients and v1.0 for servers.
## Telegraf typically acts as a client but will act as a server for listening plugins that use HTTP.
# tls_min_version = ""
```

Like with other directives, I can see it being useful for individual plugins to be able to override the agent default. This will allow people to say, set the default to 1.2 and then override in an individual plugin to use 1.1 for some old server they are in the process of upgrading.

File: plugins/common/tls/config.go
Comment (jdstrand) [reply plugins/common/tls/config.go:91]: In https://github.com/influxdata/telegraf/issues/8171#issuecomment-1183270132 I mentioned that 1.2 is used as a client and 1.0 is used as a server. This means that today, telegraf is enforcing 1.2 for all plugins that are acting as HTTP clients. I wonder if it makes sense to consider defaulting to 1.2 in telegraf 1.x? There is regression potential for listener plugins that would be acting as servers and defaulting to 1.0 as well as for people who might be working around the newer golang and choosing to compile telegraf with an older golang. The number of users relying on 1.0/1.1 in their telegraf listener or compiling with an older golang seems low... not to mention this PR provides the necessary escape hatch.

Fine to wait to set the default to 1.2 in 2.x, but I hadn't considered the recent Golang changes and that telegraf has already effectively changed its default for the vast majority of its users already.

File: plugins/inputs/gnmi/README.md
Comment (jdstrand) [reply plugins/inputs/gnmi/README.md:36]: I can see arguments on both sides, but what if we adjusted the global `[agent]` configuration and have it apply to all plugins?
```
[agent]
...
## Minimal TLS version to use for plugins that support TLS.
## If unset, the value defaults to Golang's default which is TLS v1.2 for clients and v1.0
## for servers. Telegraf typically acts as a client but will act as a server for listening
## plugins that use HTTP.
# tls_min_version = ""
```

Like other directives, users could then set `tls_min_version` in individual plugins which would allow using 1.2 everywhere except, for example, that one plugin that is talking to a 1.1 server (so the user gains the benefit of min 1.2 while in the process of upgrading that 1.1 server).

Comment (srebhan) [reply plugins/inputs/gnmi/README.md:36]: I thought about having a global option, but that has some issues as the minimum TLS version can also be set for servers (e.g for `http_listener_v2`) and this causes some ambiguity.

> Like other directives, users could then set tls_min_version in individual plugins which would allow using 1.2 everywhere except, for example, that one plugin that is talking to a 1.1 server (so the user gains the benefit of min 1.2 while in the process of upgrading that 1.1 server).

Well that is the case currently. Golang defaults to TLS v1.2 if you do not state otherwise, so we only deviate from v1.2 if explicitly set by the user.

File: plugins/common/tls/config.go
Comment (srebhan) [reply plugins/common/tls/config.go:91]: Hmmm maybe I'm missing something, but as we do not set the TLS minimum version up-to-now, we default to Golang's default which is TLS v1.2 since go 1.18.
What is your suggestion here?

Comment (jdstrand) [reply plugins/common/tls/config.go:91]: v1.2 is the minimum default for only clients. v1.0 is the minimum for servers (https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/crypto/tls/common.go;l=685). You're right that most of the code today is getting the 1.2 minimum version (since most of the plugins are operating as tls client), but there is some code operating as a server. My suggestion (or rather pondering) is since the upgrade to 1.18 appeared to go fine, perhaps pull the trigger in telegraf 1.x for setting 1.2 explicitly so we cover our tls server code too.

File: plugins/inputs/http/README.md
Comment (jdstrand) [inline plugins/inputs/http/README.md:54]: s/undet/unset/

File: plugins/inputs/gnmi/README.md
Comment (jdstrand) [reply plugins/inputs/gnmi/README.md:36]: > I thought about having a global option, but that has some issues as the minimum TLS version can also be set for servers (e.g for `http_listener_v2`) and this causes some ambiguity.

Where is this limitation? It shouldn't be a Golang limitation as https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/crypto/tls/common.go;l=685 says:

```
	// MinVersion contains the minimum TLS version that is acceptable.
	//
	// By default, TLS 1.2 is currently used as the minimum when acting as a
	// client, and TLS 1.0 when acting as a server. TLS 1.0 is the minimum
	// supported by this package, both as a client and as a server.
	//
	// The client-side default can temporarily be reverted to TLS 1.0 by
	// including the value "x509sha1=1" in the GODEBUG environment variable.
	// Note that this option will be removed in Go 1.19 (but it will still be
	// possible to set this field to VersionTLS10 explicitly).
```

This means that we should be able to set clients to TLS v1.1 and higher. I agree that trying to also set GODEBUG to enable v1.0 in clients is pointless since that is going away in 1.19 (I missed this 1.19 detail before). Due note that the option is also useful for hardening and requiring v1.3 ahead of Golang defaults (this is considered industry best practice but I agree is not the correct default for telegraf at this time). 

> > Like other directives, users could then set tls_min_version in individual plugins which would allow using 1.2 everywhere except, for example, that one plugin that is talking to a 1.1 server (so the user gains the benefit of min 1.2 while in the process of upgrading that 1.1 server).
> 
> Well that is the case currently. Golang defaults to TLS v1.2 if you do not state otherwise, so we only deviate from v1.2 if explicitly set by the user.

As mentioned elsewhere, v1.2 is only the Golang default when acting as a client and as you pointed out, `http_listener_v2` is when acting as a server. As such, we still have a low security issue for when telegraf is acting as a server and using a too low minimum version.

My suggestion is about being clear and explicitly setting v1.2 in the global section which:
* addresses the too-low Golang default of v1.0 for servers by setting it to v1.2
* makes it discoverable to users what telegraf is doing (ie, they don't have to know that Golang 1.18 changed the client default and that telegraf is using 1.18; they just see it in the config)
* conveniently allows users to adjust the default for all plugins to use v1.3 to harden their configuration and meet certain certification/operational requirements

As for ambiguity of server vs client telegraf default in the global section, I think we can make things clear for users with appropriate documentation. My previous suggestion as not is good as it could be. Perhaps:

```
[agent]
...
## Minimal TLS version to use for plugins that support TLS. Please
## choose one of: TLS13, TLS12 or TLS11. TLS12 is recommended for widest
## secure compatibility. Note that telegraf typically acts as a client
## but will act as a server for listening plugins that use HTTP. TLS10 is
## only supported when acting as a server, so if TLS10 is needed, please
## set in the individual plugin.
## 
tls_min_version = "TLS12"
```

(this implies that only TLS11 and higher is supported in the global section with support for setting TLS10 in the listeners' configuration. If you'd prefer to allow setting TLS10 in the global section, the comment should be adjusted that clients will be upgraded to TLS11).

Comment (srebhan) [reply plugins/inputs/gnmi/README.md:36]: My point is not on limitations it's about the ambiguity of having __one__ `tls_min_version`. As you suggest you can make it apply to both client and server side, but that would break backward compatibility for all server sides at least.

So my fear is the following scenario: People have some clients and some servers in their config and for whatever reason are constraint to < v1.2 on the server side. Instead of changing the setting in individual plugins, they will much likely change the global option to say v1.0 and be more insecure than before... I'd rather prefer to set v1.2 as a default _without_ an option, _forcing_ people to change it in the individual plugins and use the innate laziness.. ;-)

File: plugins/common/tls/config.go
Comment (srebhan) [reply plugins/common/tls/config.go:94]: Pushed for making v1.2 the (explicit) default for both server and client. Allowing to revert on a plugin basis...

File: plugins/inputs/gnmi/README.md
Comment (srebhan) [reply plugins/inputs/gnmi/README.md]: Pushed for making v1.2 the (explicit) default for both server and client. Allowing to revert on a plugin basis to mitigate the risk of users lowering the global options out of laziness...

**Comment by srebhan** [comment]
Made v1.2 the new explicit default for server and client connections. Users that require lower version can revert the used version on a per-plugin basis using `tls_min_version`. Please check the `Breaking changes` documentation in the Changelog. This should apply for the release containing this change (i.e. v1.24.0 IIRC)...

**Review by powersj — COMMENTED** [review]
Thanks for driving this.

My only concern is around a user knowing what to put in for the `tls_min_version` value. My naive assumption reading the changelog was they would put "1.0" or "1.2" Not the "TLS12" I later saw in the READMEs and sample configs.

**Comment by srebhan** [comment]
Good point. I just reused the values available for [`tlsVersionMap` keys](https://github.com/influxdata/telegraf/blob/af43d0183c4b4035d6f5cec20ff7e5683ef4149f/plugins/common/tls/common.go#L5) that already exist. Should I enumerate all available values or do you have another idea on how to describe this better?

**Comment by srebhan** [comment]
Another option would be to add `v1.0` etc to the keys linked above or print them on error...

**Comment by powersj** [comment]
I like the printing of the available options on error. It makes it very clear what versions we recognize. 

Thanks!

**Review by powersj — APPROVED** [review]
+1 - thank you, before merging we do want a +1 from security

**Comment by srebhan** [comment]
Thanks @powersj. I agree, will wait for @jdstrand to do his magic once he has some time... ;-)

**Review by jdstrand — APPROVED** [review]
This looks great! I particularly like the tests that do a real connection based on telegraf's `ClientConfig` and `ServerConfig` structs (it made me wonder if we could start up an http_listener as the server and then use some other input plugin as the client, but that isn't a blocker for this PR).

In addition to the PR review, I manually tested `tls_min_version` with both `inputs.http` (client) and `inputs.http_listener_v2` (server) and it worked as expected.

File: plugins/common/tls/config.go
Comment (srebhan) [reply plugins/common/tls/config.go:87]: added

File: plugins/common/tls/config.go
Comment (srebhan) [reply plugins/common/tls/config.go:146]: added

**Comment by srebhan** [comment]
@powersj and @jdstrand, may I ask for another approval?

**Review by jdstrand — APPROVED** [review]
Thanks for the updates!

**Comment by srebhan** [comment]
@jdstrand that's because I manually put the "breaking change" documentation to the Changelog. Not sure how to do it with a happy linter. :-)

**Review by powersj — APPROVED** [review]

**Comment by mohsin106** [comment]
Hi,
I'm using the Telegraf version 1.23.4-alpine Docker image and still experiencing an issue connecting when setting the `tls_max_version` parameter.
Error:
```2022-08-26T15:15:54Z E! [telegraf] Error running agent: Error loading config file /etc/telegraf/telegraf.conf: plugin inputs.jti_openconfig_telemetry: line 14: configuration specified the fields ["tls_max_version"], but they weren't used```

This is my input JTI stanza:
```
[[inputs.jti_openconfig_telemetry]]
  servers = ["router1.mgt.net:50051"]
  sensors = ["interfaces /interfaces/interface","lldp /lldp","optics /junos/system/linecard/optics/"] 
  sample_frequency = "60000ms"
  username = "$user"
  password = "$password"
  client_id = "$containerName"
  enable_tls = true
  tls_ca = "/etc/telegraf/router_ca.pem"
  insecure_skip_verify = false
  tls_max_version = "TLS12"
```

Also, tried putting the `tls_max_version` under the `agent` settings:
```
[agent]
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 15000
  flush_interval = "10s"
  debug = false
  quiet = false
  tls_max_version = "TLS12"
  hostname = "lab-ptx-deployment"
  omit_hostname = false
```
But get same error:
```
2022-08-26T15:25:14Z E! [telegraf] Error running agent: Error loading config file /etc/telegraf/telegraf.conf: line 1: configuration specified the fields ["tls_max_version"], but they weren't used
```
telegraf-alpine:1.13.3 continues to work, anything newer gives the above errors.

**Pull request merged**
